FSI application upgrade

rubenzorrilla opened the pull request:
Multiple developments in the FSI application:
- Initial MPI developments
- FSI MPI tests creation
- Updating previous tests 
- Improving the partitioned FSI utilities
- Creation of a nodal update utility

jcotela committed:
Adding MPI test suites (not working yet)

jcotela committed:
Initial commit of a new test for the MPI version of the Aitken accelerator

jcotela committed:
Adding more tests.

jcotela committed:
Adding MPI version of Aitken to factory.

jcotela committed:
Moving FSI variables to core (so that we can use them from trilinos).

jcotela committed:
Fixing compilation warning.

jcotela committed:
Adding missing Register for FSI variables.

jcotela committed:
Adding SetValue to ublas space (and making it static in TrilinosSpace).

jcotela committed:
Adding tests to FSI.

jcotela committed:
Adding files for tests.

jcotela committed:
Moving FSI variables to core.

jcotela committed:
Renaming test.

jcotela committed:
moved test files.

jcotela committed:
resolving merge conflicts.

jcotela committed:
Adding back variables lost in merge conflict.

jcotela committed:
Adding functions that were only implemented for trilinos space.

jcotela committed:
Ongoing work: adding functions to operate with Epetra vectors.

jcotela committed:
Ongoing work on MPI Aitken accelerator.

jcotela committed:
Adding script to launch mpi tests from command line.

jcotela committed:
Finalizing implementation of MPI Aitken.

rubenzorrilla committed:
Merge branch 'master'

Conflicts:
	applications/trilinos_application/trilinos_space.h
	kratos/CMakeLists.txt
	kratos/includes/kratos_application.h
	kratos/python/add_containers_to_python.cpp
	kratos/sources/variables.cpp

rubenzorrilla committed:
Fixing merging and OpenMP errors

rubenzorrilla committed:
Adding mpi.size check

At the moment, the mpi convergence accelerator tests must be run with 2 processes. Adding a security check to avoid number of processes different than 2.

rubenzorrilla committed:
Creation of a Trilinos partitioned FSI base solver

Scratch base solver for partitioned FSI problems (still under development).

rubenzorrilla committed:
Advances in Trilinos FSI Python solvers

rubenzorrilla committed:
Advances in Trilinos FSI Python solvers

rubenzorrilla committed:
Merge remote-tracking branch 'refs/remotes/origin/feature-fsi-mpi'

Conflicts:
	applications/FSIapplication/python_scripts/trilinos_partitioned_fsi_dirichlet_neumann_solver.py

rubenzorrilla committed:
Merged branch master into feature-fsi-mpi

rubenzorrilla committed:
Merged branch master into feature-fsi-mpi

rubenzorrilla committed:
Ensure that block builder and solver is used in FSI

Specially for the Trilinos case, in where the ComputeReactions() method is not implemented yet for the Elimination builder and solver.

rubenzorrilla committed:
More advances

rubenzorrilla committed:
Merged branch master into feature-fsi-mpi

rubenzorrilla committed:
Merged branch master into feature-fsi-mpi

rubenzorrilla committed:
Moving interface correction to partitioned utilities

rubenzorrilla committed:
Merged branch master into feature-fsi-mpi

rubenzorrilla committed:
Exporting more Trilinos FSI partitioned utilities methods

rubenzorrilla committed:
More advances

rubenzorrilla committed:
Minor corrections

Changes in serial base solver according to the MPI modifications

rubenzorrilla committed:
More advances

rubenzorrilla committed:
Merged branch master into feature-fsi-mpi

rubenzorrilla committed:
Minor correction and clean up

rubenzorrilla committed:
Minor correction and clean up

rubenzorrilla committed:
Merged branch feature-fsi-mpi into feature-fsi-mpi

rubenzorrilla committed:
Merged branch master into feature-fsi-mpi

rubenzorrilla committed:
Minor correction in spring tests

rubenzorrilla committed:
Updating partitioned FSI utilities

rubenzorrilla committed:
Updating serial solver according to the new modifications in partitioned utils.

rubenzorrilla committed:
Minor modification un residual vector update

rubenzorrilla committed:
Merged branch master into feature-fsi-mpi

rubenzorrilla committed:
Merged branch master into feature-fsi-mpi

rubenzorrilla committed:
Structure Neumann BC MPI set up

rubenzorrilla committed:
Addition of norms computation

These two methods allow to compute some variables norms in both the fluid and the structure interfaces. Note that its usage is intended for debugging purposes.

rubenzorrilla committed:
Python FSI solvers updated and cleaned up

rubenzorrilla committed:
Adition of debugging utilities to fsi_partitioned_utilities

rubenzorrilla committed:
Removed ComputeCorrectedInterfaceDisplacementDerivatives() method

Now the tasks that were carried out by this method are supposed to be performed by the nodal_update_utilities.

rubenzorrilla committed:
Addition of nodal_update_utilities

rubenzorrilla committed:
Updating Python exportation according to the last changes

rubenzorrilla committed:
Creation of the MESH_ACCELERATION variable

rubenzorrilla committed:
Trilinos FSI Python solvers update and clean up

rubenzorrilla committed:
Merged branch master into feature-fsi-mpi

rubenzorrilla committed:
Merge branch 'master'

Conflicts:
	applications/FSIapplication/tests/KratosExecuteFSIProblemEmulatorTest.py

rubenzorrilla committed:
Using the structural Python solvers wrapper

rubenzorrilla committed:
Merge branch 'master'

Conflicts:
	applications/FSIapplication/custom_python/add_custom_utilities_to_python.cpp
	applications/FSIapplication/python_scripts/partitioned_fsi_base_solver.py
	applications/FSIapplication/tests/test_FSIApplication.py
	kratos/python/add_cfd_variables_to_python.cpp
	kratos/python/add_cfd_variables_to_python.h

rubenzorrilla committed:
Updating defaults according to mesh solvers

rubenzorrilla committed:
Removing debug prints

rubenzorrilla committed:
Minor clean up

rubenzorrilla committed:
Updating NonConformantOneSideMap tests

The tests corresponding to the NonConformantOneSideMap have been updated to be less spurious and simple.

rubenzorrilla committed:
Mok benchmark validation test minor corrections

rubenzorrilla committed:
Updating FSIProblemEmulatorTest

The tests corresponding to the FSIProblemEmulatorTest have been updated to be less spurious and simple format.

rubenzorrilla committed:
Updating Mok benchmark test

The Mok benchmark test has been updated to be less spurious and simple format.

rubenzorrilla committed:
Updating the rest of tests

Updating the rest of tests to the current naming convention

RiccardoRossi commented on applications/FSIapplication/custom_utilities/nodal_update_utilities.h:
```diff
@@ -0,0 +1,395 @@
+            const array_1d<double, 3>& vmesh_n = it_node->FastGetSolutionStepValue(MESH_VELOCITY, 1);      // Previous step mesh velocity
+            const array_1d<double, 3>& amesh_n = it_node->FastGetSolutionStepValue(MESH_ACCELERATION, 1);  // Previous step mesh acceleration
+
+            const array_1d<double, 3>& umesh_n1 = it_node->FastGetSolutionStepValue(MESH_DISPLACEMENT, 1); // Current step mesh displacement
```
this is accessing to the old step, not to the current one

rubenzorrilla replied:
My bad. Copy & Paste error... Thank you!

RiccardoRossi commented on applications/FSIapplication/custom_utilities/partitioned_fsi_utilities.hpp:
```diff
@@ -120,81 +128,39 @@ class PartitionedFSIUtilities
-        if (FixVariable)
-        {
-            // Get the variable components
-            typedef VariableComponent< VectorComponentAdaptor<array_1d<double, 3> > > component_type;
```
@pooyan-dadvand , we should find a good way of doing this. (@ruben please ignore this comment, is not for you)

pooyan-dadvand replied:
I agree, Actually I want to remove the components in the new data structure and use variable<double> instead with some additional information in the key.

RiccardoRossi commented on applications/FSIapplication/custom_utilities/partitioned_fsi_utilities.hpp:
```diff
@@ -247,94 +215,245 @@ class PartitionedFSIUtilities
-    /*@} */
-    /**@name Protected Operations*/
-    /*@{ */
+            ux_norm += std::pow(it_node->FastGetSolutionStepValue(DISPLACEMENT_X), 2);
```
the procedure you used accesses 3 times the database.
inestead it would be better to do

    const array_1d<double,3>& d = it_node->FastGetSolutionStepValue(DISPLACEMENT)
    ux_norm += d[0]*d[0];
    uy_norm += d[1]*d[1];
    ...

rubenzorrilla replied:
Done!

RiccardoRossi commented on applications/FSIapplication/custom_utilities/partitioned_fsi_utilities.hpp:
```diff
@@ -247,94 +215,245 @@ class PartitionedFSIUtilities
+        for(int k=0; k<static_cast<int>(rLocalMesh.NumberOfNodes()); ++k)
+        {
+            const ModelPart::NodeIterator it_node = local_mesh_nodes_begin+k;
+            const array_1d<double, 3> disp = it_node->FastGetSolutionStepValue(MESH_DISPLACEMENT);
```
get a reference not a copy.

rubenzorrilla replied:
Done!

RiccardoRossi commented on applications/FSIapplication/custom_utilities/partitioned_fsi_utilities.hpp:
```diff
@@ -247,94 +215,245 @@ class PartitionedFSIUtilities
+        for(int k=0; k<static_cast<int>(rLocalMesh.NumberOfNodes()); ++k)
+        {
+            const ModelPart::NodeIterator it_node = local_mesh_nodes_begin+k;
+            const array_1d<double, 3> disp = it_node->FastGetSolutionStepValue(DISPLACEMENT);
```
get a reference not a copy

rubenzorrilla replied:
Done!

RiccardoRossi commented on applications/FSIapplication/python_scripts/convergence_accelerator_factory.py:
```diff
@@ -2,52 +2,73 @@
+        configuration.ValidateAndAssignDefaults(_aitken_defaults)
+
         if (configuration["acceleration_type"].GetString() == "Aitken"):
-            convergence_accelerator = KratosFSI.AitkenConvergenceAccelerator(configuration["w_0"].GetDouble())       
```
why don't you pass the parameters in the constructor? it is designed for this, and it will allow you simplifying a lot this factory.

RiccardoRossi replied:
note that the defaults and the validations should be done within the constructor of each separate accelerator

rubenzorrilla replied:
You are right, it can be done just creating a new constructor with a unique Parameters object as argument. The point is how we can firstly perform the ValidateAndAssignDefaults() in case there is a const member variable...

RiccardoRossi replied:
I would say in extreme.case you can use a const_cast.

In any case the advantage of having everything in the constructor is large, for many reasons.

rubenzorrilla replied:
Done!

RiccardoRossi reviewed (changes requested):
small stuff here and there. 

passing parameters in the constructor is the most important change

rubenzorrilla committed:
Minor correction in current step velocity and acceleration computation

Copy & Paste error correction in the computation of the corrected current step velocity and acceleration computation.

rubenzorrilla committed:
Minor review improvements

rubenzorrilla committed:
Merged master into feature-fsi-mpi

rubenzorrilla committed:
Convergence accelerators new constructor

A new constructor that takes a parameters object as unique argument has been created for all the FSI convergence accelerators.

RiccardoRossi reviewed (approved):
looks correct to me
